### PR TITLE
harmonization and clarification of dice losses variants docs and associated tests

### PR DESCRIPTION
**Pull request by Lucas-rbnt**
### Description

This PR aims to clarify and harmonise the code for the DiceLoss variants in the `monai/losses/dice.py` file. With the `to_onehot_y` `softmax` and `sigmoid` arguments, I didn't necessarily understand the ValueError that occurred when I passed a target of size NH[WD]. I had a bit of trouble reading the documentation and understanding it. I thought that they had to be the same shape as they are displayed, unlike the number of dimensions in the input, so I added that.
Besides, in the documentation is written:
```python
"""
raises:
      ValueError: When number of channels for target is neither 1 nor the same as input.

"""
```
Trying to reproduce this, we give an input with a number of channels $N$ and target a number of channels of $M$, with $M \neq N$ and $M > 1$.
```python
loss = DiceCELoss()
input = torch.rand(1, 4, 3, 3)
target = torch.randn(1, 2, 3, 3)
loss(input, target)
>: AssertionError: ground truth has different shape (torch.Size([1, 2, 3, 3])) from input (torch.Size([1, 4, 3, 3]))
```
This error in the Dice is an `AssertionError` and not a `ValueError` as expected and the explanation can be confusing and doesn't give a clear idea of the error here. The classes concerned and harmonised are `DiceFocalLoss`, `DiceCELoss` and `GeneralizedDiceFocalLoss` with the addition of tests that behave correctly and handle this harmonisation. 

Also, feel free to modify or make suggestions regarding the changes made in the docstring to make them more understandable (in my opinion, but other readers and users will probably have a different view).

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/losses/dice.py [hunk 1]
```diff
@@ -778,12 +778,22 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
 
         Raises:
             ValueError: When number of dimensions for input and target are different.
-            ValueError: When number of channels for target is neither 1 nor the same as input.
+            ValueError: When the target channel isn't either one-hot encoded or categorical with the same shape of the input.
```
Comment (KumoLiu) [inline monai/losses/dice.py:781]: The error message seems to indicate that the data must either be in one-hot format or consistent with the original image, which appears to slightly differ from the initial implication.

File: monai/losses/dice.py [hunk 2]
```diff
@@ -1038,3 +1066,11 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
 generalized_dice_focal = GeneralizedDiceFocalLoss
 generalized_wasserstein_dice = GeneralizedWassersteinDiceLoss
+
+if __name__ == "__main__":
+    loss = DiceFocalLoss(to_onehot_y=False, softmax=True)
+    input = torch.randn(1, 3, 128, 128, 128)
+    target = torch.randn(1, 4, 128, 128, 128)
+    print(input.dim())
+    print(target.dim())
+    loss(input, target)
```
Comment (KumoLiu) [inline monai/losses/dice.py:1076]:
```suggestion
```

**Review by KumoLiu — COMMENTED** [review]
Thank you for this pull request, overall, the updates are looking good to me. However, I do have two minor comments inline for your consideration.

File: monai/losses/dice.py (hunk 1 above)
Comment (Lucas-rbnt) [reply monai/losses/dice.py]: Yes, sorry, on re-reading the original message is clearer, I've changed that, hoping it's better now!

**Review by KumoLiu — APPROVED** [review]

**Pull request merged**